teps/0081-add-chains-subcommand-to-the-cli.md

```diff
@@ -0,0 +1,326 @@
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This proposal aims at treating `Chains` common actions as first class citizens
+by adding a new command to the Tekton CLI.
+
+## Motivation
+
+When working with `Chains`, the user experienced will be enhanced if users are
+able to
+*  use the Tekton CLI rather than a combination of other tools like `kubectl`,
+`jq`, `base64`, etc.
+* not having to memorize which exact TaskRun annotation or key to query.
+* discover available `Chains` features via the CLI and/or its documentation.
+
+### Goals
+
+Add the ability to run `tkn chains <chains_action>` fom the CLI.
+
+### Non-Goals
+
+While the idea is to use the Tekton CLI to replace a combination of shell
+commands, we may not want to implement all exotic variations and focus only on
+the most common use cases in order to avoid overloading the `chains` command.
+
+### Use Cases
+
+#### Use Case 1
+
+Extract the Chains payload from a taskrun:
+
+```bash
+$ tkn chains payload cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v
+{
+  "conditions": [
+    {
+      "type": "Succeeded",
+      "status": "True",
+      "lastTransitionTime": "2021-08-31T15:19:45Z",
+      "reason": "Succeeded",
+      "message": "All Steps have completed executing"
+    }
+  ],
+  "podName": "cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v-pod-qptxn",
+  "startTime": "2021-08-31T15:19:17Z",
+  "completionTime": "2021-08-31T15:19:45Z",
+  ...
+}
+```
+
+Instead of:
+
+```bash
+TASKRUN=cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v
+TASKRUN_UID=$(kubectl get taskrun $TASKRUN -o=json | jq -r '.metadata.uid')
+kubectl get taskrun $TASKRUN -o=json \
+  | jq -r ".metadata.annotations[\"chains.tekton.dev/payload-taskrun-$TASKRUN_UID\"]" \
+  |base64 --decode \
+  | jq
+```
+
+#### Use Case 2
+
+Extract the signature from a taskrun:
+
+```bash
+$ tkn chains signature cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v
+MEUCIQCHS4in92kQW6B+LfnEilpRcOg5akspPh4mQ+tSaEq/jgIgdl83eCyw41xFC8xti6j0/TgzXkKVixfD30yenabWyHU=
+```
+
+Instead of:
+
+```bash
+TASKRUN=cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v
+TASKRUN_UID=$(kubectl get taskrun $TASKRUN -o=json | jq -r '.metadata.uid')
+kubectl get taskrun $TASKRUN -o=json \
+  | jq -r ".metadata.annotations[\"chains.tekton.dev/signature-taskrun-$TASKRUN_UID\"]"
+```
+
+#### Use Case 3
+
+Change the provenance format:
+
+```bash
+$ tkn chains format in-toto
```

Comment:
Should we have a subcommand for *configuring* chains (in OpenShift we have `oc adm` for example). `tkn chains adm …` or even `tkn adm chains …` (which would mean we could/would also introduce similar features for other component such as pipeline, triggers, …)

Reply:
@priyawadhwa proposed `tkn chains config` which would do it.

We could add a "convention" stating that each command which would offer the ability to edit its related configuration must provide a `config` sub-command. That's kind of the reverse of what you are proposing, where all configuration sub-commands would be grouped under `tkn adm`. Both approaches make sense to me, although I like the config `tkn <command> config` idea a bit more.

Reply:
@rgreinho the only "trick" with `tkn <command> config` is that, it applies to `chains` well, but not for the rest, like `tkn task config` or `tkn pipeline config`, what does it do ? `pipeline`, `task`, `clustertask`, `pipelinerun` and `taskrun` are all subcommand that are "configured" together (`tektoncd/pipeline`). This is similar for the triggers subcommand for example.

Reply:
Ah I see! Then a `tkn adm chains` makes more sense after this explanation 👍